docs/architecture-and-design.md

```diff
@@ -394,3 +400,62 @@ In order for a robot to move to the desired destination of a `MoveIntent`, the N
+
+## Simulated Integration Tests Architecture
+Fortunately, our abstractions and [Observer system](#observer-design-pattern) make it very easy to modify our system to support simulated integration tests. There are only 2 major changes we make to the system:
+1. Implement a [Backend](#backend) to handle the simulation
+2. Add another Observer to the system that will handle the "validation"
+
+The [SimulatorBackend](#simulator-backend) and [World State Validator](#world-state-validator) implement each of these changes respectively, and are further described in their own sections.
+
+### Simulator Backend
+The `SimulatorBackend` is simply another implementation of the [Backend](#backend) interface. It uses a physics library to simulate the various components of the [World](#world), like the [Ball](#ball) and [Robots](#robot). Like any [Backend](#backend), the `SimulatorBackend` publishes the state of the [World](#world) periodically.
+
+In order to achieve determinism, the `SimulatorBackend` will publish new [World](#world)'s with a fixed time increment (in [World](#world) time), and will wait to receive [Primitives](#primitives) before simulating and publishing the next [World](#world). This means that no matter how much faster or slower the simulation runs than the rest of the system, everyting will always happen "at the same speed" from the POV of the rest of the system, since each newly published [World](#world) will be a fixed amount of time newer than the last. See the section on [Component Connections and Determinism](#component-connections-and-determinism) for why this is important.
+
+### World State Validator
+The `WorldStateValidator` is an Observer whose purpose is to check that the state of the world is "correct" according to some user-provided metrics, or is changing as expected. This is effectively the "assert" statements of our tests.
+
+There are generally 2 "types" of conditions we want to validate.
+1. Some sequence of states or actions occur **in a given order**.
+    * Eg. A robot moves to point A, then point B, then kicks the  ball
+2. Some condition is met for the duration of the test, or for "all time"
+    * Eg. The ball never leaves the field, or robots never collide
+
+We use `ValidationFunctions` to validate both types of conditions. `ValidationFunctions` are essentially functions that contain [Google Test](https://github.com/google/googletest) `ASSERT` statements, and use [Coroutines](#coroutines) to maintain state. This lets us write individual functions that can be continuously run to validate both types of conditions above.
+
+Benefits of `ValidationFunctions`:
+1. They are reuseable. We can write a function that validates the ball never leaves the field, and use it for multiple tests.
+2. They can represent assertions for more complex behavior. For example, we can build up simpler `ValidationFunctions` until we have a single `Validation` function for a specific [Tactic](#tactics).
```

Comment:
Can we nest validation functions like this? Just not sure how this works from a technical standpoint from what I remember of the design.

Reply:
Yes they could be nested (just make sure you don't nest `Continuous` functions). Should nesting be mentioned explicitly?

Reply:
Nah I think this is fine.